[vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42614 [vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)**

Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

Docker image rebuild:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6700311
Vulkan Linux Build:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6701604
Vulkan Linux Test:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6703026

Differential Revision: [D23174038](https://our.internmc.facebook.com/intern/diff/D23174038)


ci-all check: https://github.com/pytorch/pytorch/pull/43364

IvanKobzarev committed:
[WIP] VulkanSdk and SwiftShader to Android docker image

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[WIP] VulkanSdk and SwiftShader to Android docker image"


Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)"


Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

[ghstack-poisoned]

ezyang commented:
@seemethere looks like another long compilation step for docker builders getting added again...

ezyang commented on aten/src/ATen/test/vulkan_test.cpp:
```diff
@@ -126,10 +126,11 @@ TEST(VulkanTest, conv2d) {
   auto t_in = at::rand({1, C, H, W}, at::device(at::kCPU).dtype(at::kFloat));
   auto t_w = at::rand({OC, C, KH, KW}, at::device(at::kCPU).dtype(at::kFloat));
   auto t_b = at::zeros({OC}, at::device(at::kCPU).dtype(at::kFloat));
-  auto stride = c10::IntArrayRef{1};
```
Oof, this definitely shouldn't have been written this way haha

ezyang commented:
This looks plausible. I'll let other people take a look.

IvanKobzarev commented:
> @seemethere looks like another long compilation step for docker builders getting added again...

Yes, swiftshader build includes building llvm-10 that takes ~30 minutes with current configuration, but if to change docker build job's circleci `resource-class` to `2xlarge` (currently it is `large`) - It takes about ~15mins.

seemethere commented:
Is swiftshader something that could theoretically be built once and then used by multiple things?

We recently just got rid of the `llvm` 9 build that we had in the docker builds since it only really needed to be built once. https://github.com/pytorch/pytorch/pull/43038

Perhaps we can take a similar approach to `swiftshader`

IvanKobzarev commented:
> Is swiftshader something that could theoretically be built once and then used by multiple things?
> 
> We recently just got rid of the `llvm` 9 build that we had in the docker builds since it only really needed to be built once. #43038
> 
> Perhaps we can take a similar approach to `swiftshader`

I will try to do the same with swiftshader.

IvanKobzarev committed:
Update on "[vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)"


Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

Docker image rebuild:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6700311
Vulkan Linux Build:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6701604
Vulkan Linux Test:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6703026

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)"


Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

Docker image rebuild:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6700311
Vulkan Linux Build:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6701604
Vulkan Linux Test:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6703026

[ghstack-poisoned]

IvanKobzarev commented:
> Is swiftshader something that could theoretically be built once and then used by multiple things?
> 
> We recently just got rid of the `llvm` 9 build that we had in the docker builds since it only really needed to be built once. #43038
> 
> Perhaps we can take a similar approach to `swiftshader`

@seemethere,

I put prebuilt swiftshaders to aws (it is about 900mb) and unpacking of it instead of building works fine, tests passed.

Is it ok to leave it on aws and download it from there instead of repeating the whole setup of llvm built on docker image in pytoch/builder?

seemethere commented:
> > Is swiftshader something that could theoretically be built once and then used by multiple things?
> > We recently just got rid of the `llvm` 9 build that we had in the docker builds since it only really needed to be built once. #43038
> > Perhaps we can take a similar approach to `swiftshader`
> 
> @seemethere,
> 
> I put prebuilt swiftshaders to aws (it is about 900mb) and unpacking of it instead of building works fine, tests passed.
> 
> Is it ok to leave it on aws and download it from there instead of repeating the whole setup of llvm built on docker image in pytoch/builder?

That should be fine as well as long as we have a way to recreate them if need be.

IvanKobzarev committed:
Update on "[vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)"


Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

Docker image rebuild:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6700311
Vulkan Linux Build:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6701604
Vulkan Linux Test:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6703026

Differential Revision: [D23174038](https://our.internmc.facebook.com/intern/diff/D23174038)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)"


Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

Docker image rebuild:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6700311
Vulkan Linux Build:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6701604
Vulkan Linux Test:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6703026

Differential Revision: [D23174038](https://our.internmc.facebook.com/intern/diff/D23174038)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan][ci] Vulkan tests running on linux build via swiftshader (added to docker)"


Vulkan backend linux build (USE_VULKAN=1) and running Vulkan tests using software Vulkan implementation via [swiftshader](https://github.com/google/swiftshader)

Vulkan linux build needs VulkanSdk and running tests needs Swiftshader.
swiftshader needs to be compiled using clang toolchain, added them to bionic-clang-9 docker image.


VulkanSdk will be downloaded from aws;
Swiftshader is cloned from github, as it has many submodules , commit hash is fixed in install_swiftshader script.

To pass all the tests:
Disabled adaptive_avg_pool2d_2 as it needs at::view which will be landed in https://github.com/pytorch/pytorch/pull/42676 and after that can be enabled

Change strides, padding, dilation params in tests to vector

Docker image rebuild:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6700311
Vulkan Linux Build:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6701604
Vulkan Linux Test:
https://app.circleci.com/pipelines/github/pytorch/pytorch/200251/workflows/465f911f-f170-47e1-954e-b9605d91abd8/jobs/6703026

Differential Revision: [D23174038](https://our.internmc.facebook.com/intern/diff/D23174038)


ci-all check: https://github.com/pytorch/pytorch/pull/43364

[ghstack-poisoned]
